src/task_system/task_system_server/Controllers/TaskSystemController.cs

```diff
@@ -26,37 +26,41 @@ public TaskSystemController(ITaskSystemRepository taskSystemRepository, IPermiss
         [HttpGet]
         public async Task<IActionResult> GetTasks([FromQuery] QueryObject query)
         {
-            if (!Request.Headers.TryGetValue("X-API-Key", out var apiKey) || !_permissionValidator.ValidatePermissions(PermissionNames.VIEW_TASKS, apiKey!))
+            if (!Request.Headers.TryGetValue("X-API-Key", out var apiKey))
                 return Unauthorized(new TaskSystemResponse<string>(ResponseMessages.UNAUTHORIZED, null));
 
+            if(!_permissionValidator.ValidatePermissions(PermissionNames.VIEW_TASKS, apiKey) && 
+                (!query.AccountId.HasValue || !_permissionValidator.ValidateAccountId(query.AccountId.Value, apiKey!) || !_permissionValidator.ValidatePermissions(PermissionNames.CREATE_TASKS, apiKey!)))
+            {
+                return Unauthorized(new TaskSystemResponse<string>(ResponseMessages.UNAUTHORIZED, null));
+            }
+
             if (!ModelState.IsValid)
                 return BadRequest(ModelState);
-
+                
             var tasks = await _taskSystemRepository.GetTasksAsync(query);
 
             if (!tasks.Any())
-            {
                 return NotFound(new TaskSystemResponse<string>(ResponseMessages.GET_TASKS_FAILED, null));
```

Comment:
## Not actionable
I recently learned we have been using 404s incorrectly. 

The only time a 404 should be returned is if a specific resource is not found, eg, GET /tasks/1 if that task does not exist.

However, if we are getting a list of resources, (eg, GET /tasks) and there are no tasks found, we return 200 with an empty list. 

The reasoning is that the **list** of resources was found, it just happens to be empty.

Unless we want to change everything everywhere, for consistency, I recommend leaving it as is